#1918: Add documentation on how to run IDEasy locally using intellij (#1941)
devonfw/IDEasy: documentation/contributing/run-ideasy-locally.adoc

* Set Main class to `com.devonfw.tools.ide.cli.Ideasy`
* In Program arguments, add the cli arguments you want to pass to IDEasy (in the example a call to open vscode with the force-option)
* Set Working directory to the IDEasy project folder

With this, you should have a working run configuration which starts IDEasy from your local branch with the provided arguments.
When you want to test other functionality of IDEasy, change the Program arguments in the run configuration accordingly.

This shows an example run configuration for IDEasy:

image::../images/run_ide_locally_guide/run_config_example.png[]

=== Debug IDEasy in Intellij

With the just created run configuration, we can also now use the debugger in Intellij to debug IDEasy using breakpoints.

Just set breakpoints anywhere in the code and then use the "Debug" button with the run configuration.

== Run IDEasy locally from inside Visual Studio Code

Running IDEasy from inside Visual Studio Code works similarly to IntelliJ and also supports debugging with breakpoints.
It requires the https://marketplace.visualstudio.com/items?itemName=vscjava.vscode-java-pack[Extension Pack for Java] to be installed in VS Code.

> To create a launch configuration, create or open the file `.vscode/launch.json` in the root of the IDEasy project and add the following configuration:

Reviewer: ```suggestion
To create a launch configuration, open the *Run and Debug* panel (kbd:[Ctrl+Shift+D]) and select *create a launch.json file*. This will automatically create a ready-to-go launch configuration that you can modify with relevant launch arguments.

You can also manually create a launch configuration by creating and opening the file `.vscode/launch.json` in the root of the IDEasy project and adding the following configuration:
```
VSCode can also automatically create the launch.json file. I feel like that is the easier way and should therefore be explained here.

[source,json]
----
{
  "version": "0.2.0",
  "configurations": [
    {
      "type": "java",

>       "name": "IDEasy",

Reviewer: ```suggestion
      "name": "Ideasy",
```
Very minor optional change to better match the automatically generated launch config which depends on the class name (Ideasy).

      "request": "launch",
      "mainClass": "com.devonfw.tools.ide.cli.Ideasy",
      "projectName": "ide-cli",

>       "args": "install mvn",

Reviewer: ```suggestion
      "args": ["install", "mvn"],
```
Multiple arguments must be provided in a list. Otherwise, it will fail.

      "cwd": "${workspaceFolder}"
    }
  ]
}
----

The relevant settings are:

* `mainClass`: must be set to `com.devonfw.tools.ide.cli.Ideasy`
* `projectName`: must be set to `ide-cli`

> * `args`: the CLI arguments to pass to IDEasy (in the example, a call to install Maven)

Reviewer: ```suggestion
* `args`: the CLI arguments to pass to IDEasy (in the example, a call to install Maven). Multiple arguments must be passed as a list of Strings.
```

To achieve this, there are multiple options. These include running IDEasy from inside an IDE such as IntelliJ or Visual Studio Code, which also enables debugging with breakpoints and many more features.

toc::[]

== Run IDEasy locally from inside Intellij

The entrypoint for running your local changes of IDEasy is the main-method in the cli-subfolder. Since most of the functionality of IDEasy is located in the cli-submodule, this will cover nearly all the local running/testing needs.
The main-method is located in `com.devonfw.tools.ide.cli.Ideasy`. Since running the main-method without any arguments is not really useful, it is necessary to create a run configuration
in Intellij for this module/method.

The easiest way to create such a run configuration is clicking on the "run/play" button to the left of the main-method definition and then click on "Modify run configuration":

image::../images/run_ide_locally_guide/run_button_main_method.png[]

image::../images/run_ide_locally_guide/run_button_menu.png[]

In the window that is then opened, some configurations have to be made (if not set automatically already):

* Set the used SDK to the SDK of the project
* Set module to `ide-cli`

> * Set Main class to `com.devonfw.tools.ide.cli.Ideasy`

Reviewer 2: Isn't all this done automatically?
If yes, maybe change `Set ...` to `Verify that ...`

Author: yes it is done automatically when using the "play" button next to the main method, but not when creating a run configuration manually...but i also think "verify" is more  reasonable here

```suggestion
* Verify that the used SDK is the SDK of the project
* Verify that module is set to `ide-cli`
* Verify that Main class is set to `com.devonfw.tools.ide.cli.Ideasy`
```

* In Program arguments, add the cli arguments you want to pass to IDEasy (in the example a call to open vscode with the force-option)
* Set Working directory to the IDEasy project folder

With this, you should have a working run configuration which starts IDEasy from your local branch with the provided arguments.
When you want to test other functionality of IDEasy, change the Program arguments in the run configuration accordingly.

This shows an example run configuration for IDEasy:

image::../images/run_ide_locally_guide/run_config_example.png[]

=== Debug IDEasy in Intellij

With the just created run configuration, we can also now use the debugger in Intellij to debug IDEasy using breakpoints.

Just set breakpoints anywhere in the code and then use the "Debug" button with the run configuration.

== Run IDEasy locally from inside Visual Studio Code

Running IDEasy from inside Visual Studio Code works similarly to IntelliJ and also supports debugging with breakpoints.

> It requires the https://marketplace.visualstudio.com/items?itemName=vscjava.vscode-java-pack[Extension Pack for Java] to be installed in VS Code.

Reviewer 2: Isn't this installed automatically by IDEasy?
If yes, please note this so the developer knows that he has nothing to do here.

Author: Yes it is! Good catch

```suggestion
It requires the https://marketplace.visualstudio.com/items?itemName=vscjava.vscode-java-pack[Extension Pack for Java] to be installed in VS Code (This is pre-installed by IDEasy, so if you use Vs-Code from IDEasy, there is nothing to do here!).
```

> To create a launch configuration, open the *Run and Debug* panel (kbd:[Ctrl+Shift+D]) and select *create a launch.json file*. This will automatically create a ready-to-go launch configuration that you can modify with relevant launch arguments.

Reviewer 2: ```suggestion
To create a launch configuration, open the *Run and Debug* panel (press `[Ctrl][Shift][D]`) and select *create a launch.json file*.
This will automatically create a ready-to-go launch configuration that you can modify with relevant launch arguments.
```

You can also manually create a launch configuration by creating and opening the file `.vscode/launch.json` in the root of the IDEasy project and adding the following configuration:

[source,json]
----
{
  "version": "0.2.0",
  "configurations": [
    {
      "type": "java",
      "name": "Ideasy",
      "request": "launch",
      "mainClass": "com.devonfw.tools.ide.cli.Ideasy",
      "projectName": "ide-cli",
      "args": ["install", "mvn"],
      "cwd": "${workspaceFolder}"
    }
  ]
}
----

The relevant settings are:

* `mainClass`: must be set to `com.devonfw.tools.ide.cli.Ideasy`
* `projectName`: must be set to `ide-cli`
* `args`: the CLI arguments to pass to IDEasy (in the example, a call to install Maven). Multiple arguments must be passed as a list of Strings.
* `cwd`: the working directory, set to the IDEasy project root

> With this configuration saved, open the *Run and Debug* panel (kbd:[Ctrl+Shift+D]), select *IDEasy* from the dropdown, and click the run button.

Reviewer 2: ```suggestion
With this configuration saved, open the *Run and Debug* panel (press `[Ctrl][Shift][D]`), select `IDEasy` from the dropdown, and click the run button.
```

This approach is useful when you want to test your local changes without opening an IDE, or when running IDEasy in a scripted environment.

All commands below are executed from inside the `cli` subdirectory of the project.

=== Step 1: Build the JAR

Run a full Maven build to compile the project and produce the JAR:

 mvn clean install

The resulting JAR is placed in `cli/target/` and named `ide-cli-<revision>.jar`, where `<revision>` matches the value defined in `.mvn/maven.config`.

=== Step 2: Copy runtime dependencies

The JAR does not bundle its dependencies. Copy them to `target/lib/` so the JAR manifest can resolve them at runtime:

 mvn dependency:copy-dependencies -DoutputDirectory=target/lib -DincludeScope=runtime

=== Step 3: Run IDEasy

Launch IDEasy with the desired CLI arguments. Replace `<revision>` with the actual revision string (e.g. `2026.05.001-SNAPSHOT`):

 java -jar target/ide-cli-<revision>.jar <args>

For example, to install a tool:

 java -jar target/ide-cli-2026.05.001-SNAPSHOT.jar install mvn

When you want to test other functionality of IDEasy, change the CLI arguments accordingly.

> 

Reviewer 2: That is overcomplicated.
Simply do this instead:
```
mvn exec:exec -Dexec.executable=java -Dexec.args="-cp %classpath com.devonfw.tools.ide.cli.Ideasy help"
```

= Run IDEasy locally

In order to test changes you did locally in your feature branch, there is a need for running IDEasy locally from the branch with the local changes applied.

> To achieve this, there are multiple options. The first would be to run IDEasy from inside Intellij, which also enables debugging with breakpoints and many more features.

Reviewer: Your guide on how to test changes in IntelliJ is nice. Good job! However, here you state, that there are multiple options but you only list one.

As VSCode is also supported by IDEasy and many users use it, it might be good to provide similar instructions for VSCode. What do you think?

Author: Yeah this would probably be a nice addition! But I do not really use VSCode, so I would have to do some research first on how to achieve this using VsCode...maybe someone who uses VSCode regularly to contribute to IDEasy could add this documentation part.

I will also add another section on how to build/run from a local .jar file and will link to the graal-vm build guide for local native build tests.

Author: I have added instructions for VSCode aswell 👍
